[GPU][Codegen] Allowing mfma for narrow problem config sizes

jerryyin opened the pull request:
The motivation of this PR is convolution performance for resnet50 configs. With this PR (and a few pending ones), conv performance with igemm pipeline get decent speedup in situation where a standalone dimension size is smaller than intrinsic size. (Take dispatch 69 as example, the select tile m:7, n:512, k:4608 will be rejected from mfma because m tile is smaller than intrinsic size of 16). This happens because previously we are too defensive about when to use intrinsic: in situation when alignment is not required, we still enforce mfma to be picked up only when m/n/k tiles are all larger than intrinsic size.

With @nirvedhmeshram's https://github.com/iree-org/iree/pull/19271 and https://github.com/iree-org/iree/pull/19484, padding is allowed in tile and fuse matmul and igemm tile and fuse pipelines, it is no longer necessary to be as conservative as before. I am therefore getting rid of the conditional check that blocks mfma from being picked up. 

This will impact a few pipelines that use `canTargetIntrinsic()`:
 - `LLVMGPUPadAndVectorDistribute` will allow narrow m/n/k dimension sizes for batch matmul
 - In `iree-codegen-rocdl-configuration-pipeline`, will allow narrow m/n/k dimension sizes for matmul (instead of warp reduction)

jerryyin committed:
Allowing mfma for narrow problem config sizes

Signed-off-by: jerryyin <zhuoryin@amd.com>

qedawkins commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_warp_reduction.mlir:
```diff
@@ -37,45 +37,3 @@ hal.executable private @warp_reduction {
-// TODO: we probably need to revisit the configuration heuristics here.
-
-//     CHECK-LABEL: llvm.func @warp_reduction_large_vector
-// CHECK-COUNT-640:   rocdl.ds_bpermute
```
This should continue going down the warp reduction pipeline. We can add some basic checks to prevent matching small M or N cases (maybe <= 4).

jerryyin replied:
Sounds reasonable, though it'd be hard to establish a precise boundary here. I have some clarifying questions to ask:
 - I assume we don't have a split-K algorithm in a certain pipeline that will allow M or N that get stacked when computing gemm with small M or N?
 - This boundary can be hard to determine because perf will also be dependent on size of K. For the sake of resnet50, 4 should be okay. (I'll double check). Let's agree on picking 4 for now and leaving a TODO to indicate precisely deciding the threshold between whether or not picking intrinsic will require a sweep of tuning parameters.

As a side question, I'd like to understand a bit more about warp reduction used for matmul for this unit test:
 - As is, with size of [1x1280] x [1280x1280] (mk x kn), looks like each workgroup computes [1x1280] x [1280x64]. So it'd be 64 dot product, each dot product of size 1280?
 - When M is 4 and assuming each workgroup continue to compute 64 results. This is [4x1280] x [1280x64]. Does it just compute 4 times more compared with previous case?

qedawkins replied:
> TODO to indicate precisely deciding the threshold between whether or not picking intrinsic will require a sweep of tuning parameters.

TODO was exactly what I was thinking, just that at least for now we're relying on warp reduction for matvec performance in some models (llama) and can't change the defaults there without benchmarks justifying it.

> When M is 4 and assuming each workgroup continue to compute 64 results. This is [4x1280] x [1280x64]. Does it just compute 4 times more compared with previous case?

I'm not sure exactly how the configuration would work out, but my guess is that each workgroup would still compute `1x1280` x `1280x64` and we would just have 4 times as many workgroups. It's probably not the ideal solution but it's the best option we have today. If we could get better perf with MFMA though that would be great!

jerryyin replied:
Thanks! I'll flip to 4 and add TODO in the next commit. I'm going to add it for all m, n and k.

Good to get some confirmation about idea of using warp reduction in matmul. Since we are picking 4 as the cutoff, this case will be falling back for warp reduction to execute it. Maybe another day for better perf with MFMA part :-p

qedawkins commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_vector_distribute_gfx942.mlir:
```diff
@@ -259,31 +259,6 @@ func.func @unaligned_m_batch_matmul_64x72x1280x1280() {
-}
-// Check that we don't support LLVMGPUPadAndVectorDistribute for narrow N/M atm.
-// CHECK-NOT:      #iree_codegen.translation_info<pipeline = LLVMGPUPadAndVectorDistribute
-// CHECK-LABEL: func.func @narrow_n_batch_matmul_64x968x4x320_f16()
```
This test should stay because the PadAndVectorDistribute pipeline still doesn't support narrow m/n. We can add an additional configuration flag to GPUHeuristics if needed to indicate `supportsNarrowMN` or something like that.

jerryyin replied:
Nice catch. May I ask why `PadAndVectorDistribute` can't handle narrow m/n? I was assuming it can after seeing the previous unit test: `unaligned_m_batch_matmul_64x72x1280x1280` that supports unaligned case. (I have no idea about how it get lowered) But assuming it can support unaligned m case, then support this for narrow M should also work because both are about padd more zeros?

Note that the follow-up is just for my own sanity :-), to make sure I'm learning why

qedawkins replied:
IIRC it's because either padding or a later step fails in the narrow case because we end up with static sizes. If it works now that would be great though.

jerryyin replied:
I want to be extra careful about this, (since I know so little about what PadAndVectorDistribute does) so just documenting the steps I've performed below:

## Experiment

Test case (the same case I've deleted above):
```mlir
!LHS_TYPE = tensor<64x968x320xf32>
!RHS_TYPE = tensor<64x320x4xf32>
!RESULT_TYPE = tensor<64x968x4xf32>
func.func @batch_matmul(%lhs : !LHS_TYPE, %rhs : !RHS_TYPE) -> !RESULT_TYPE {
    %c0 = arith.constant 0.0 : f32
    %empty = tensor.empty() : !RESULT_TYPE
    %fill = linalg.fill ins(%c0 : f32) outs(%empty : !RESULT_TYPE) -> !RESULT_TYPE
    %mm = linalg.batch_matmul ins(%lhs, %rhs : !LHS_TYPE, !RHS_TYPE) outs(%fill : !RESULT_TYPE) -> !RESULT_TYPE
    return %mm : !RESULT_TYPE
}
```

Step 1: compile and run on cpu:
```bash
iree-compile --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-cpu=host ./test_bmm.mlir -o out_cpu.vmfb
iree-run-module --device=local-task --module=out_cpu.vmfb --input='64x968x320xf32=@64x968x320xf32.bin' --input='64x320x4xf32=@64x320x4xf32.bin' --function=batch_matmul --output=@cpu_output.npy
```

Step 2: compile and run on gpu:
```bash
iree-compile --iree-hal-target-backends=rocm --iree-hip-target=gfx942 --iree-codegen-llvmgpu-use-vector-distribution --iree-codegen-llvmgpu-use-unaligned-gemm-vector-distribution --iree-codegen-llvmgpu-use-igemm=false ./test_bmm.mlir -o out.vmfb
iree-run-module --device=hip --module=out.vmfb --input='64x968x320xf32=@64x968x320xf32.bin' --input='64x320x4xf32=@64x320x4xf32.bin' --function=batch_matmul --expected_output=@cpu_output.npy
```

Result
```bash
EXEC @batch_matmul
[SUCCESS] all function outputs matched their expected values.
```

## IR dump after `LLVMGPUSelectLoweringStrategyPass`

```mlir
// -----// IR Dump After LLVMGPUSelectLoweringStrategyPass (iree-llvmgpu-select-lowering-strategy) //----- //
module {
  func.func @batch_matmul_dispatch_0_batch_matmul_64x968x4x320_f32() attributes {translation_info = #iree_codegen.translation_info<pipeline = LLVMGPUPadAndVectorDistribute workgroup_size = [64, 1, 1] subgroup_size = 64, {}>} {
    %cst = arith.constant 0.000000e+00 : f32
    %c0 = arith.constant 0 : index
    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags("ReadOnly|Indirect") : !flow.dispatch.tensor<readonly:tensor<64x968x320xf32>>
    %1 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(1) alignment(64) offset(%c0) flags("ReadOnly|Indirect") : !flow.dispatch.tensor<readonly:tensor<64x320x4xf32>>
    %2 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, "ReadOnly|Indirect">, #hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(2) alignment(64) offset(%c0) flags(Indirect) : !flow.dispatch.tensor<writeonly:tensor<64x968x4xf32>>
    %3 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [64, 968, 320], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<64x968x320xf32>> -> tensor<64x968x320xf32>
    %4 = flow.dispatch.tensor.load %1, offsets = [0, 0, 0], sizes = [64, 320, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<64x320x4xf32>> -> tensor<64x320x4xf32>
    %5 = tensor.empty() : tensor<64x968x4xf32>
    %6 = linalg.fill ins(%cst : f32) outs(%5 : tensor<64x968x4xf32>) -> tensor<64x968x4xf32>
    %7 = linalg.batch_matmul {lowering_config = #iree_gpu.lowering_config<{mma_kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>, promote_operands = [0, 1], reduction = [0, 0, 0, 32], subgroup_m_count = 1 : i64, subgroup_n_count = 1 : i64, workgroup = [1, 16, 16, 0]}>} ins(%3, %4 : tensor<64x968x320xf32>, tensor<64x320x4xf32>) outs(%6 : tensor<64x968x4xf32>) -> tensor<64x968x4xf32>
    flow.dispatch.tensor.store %7, %2, offsets = [0, 0, 0], sizes = [64, 968, 4], strides = [1, 1, 1] : tensor<64x968x4xf32> -> !flow.dispatch.tensor<writeonly:tensor<64x968x4xf32>>
    return
  }
}
```

Looks like it does indeed used `#iree_codegen.translation_info<pipeline = LLVMGPUPadAndVectorDistribute`

## Follow-up

What do you think the right follow-up to be? It looks like I can just revert the `CHECK-NOT` to be `CHECK` instead.

qedawkins replied:
oh cool, then yeah just changing `CHECK-NOT` to `CHECK` sounds good, thanks for the investigation!

jerryyin committed:
Adding heuristic size check

This make sure tiny gemm can still be lowered through warp reduction
pipeline.

Signed-off-by: jerryyin <zhuoryin@amd.com>

qedawkins reviewed (approved):
Cool LGTM! Also I would pay close attention to the regression tests for SDXL and llama on the CI. There is some wiggle room on those tests so it might pass small regressions.
